File: uithack-2023/crypto/oracle/middleman.js
const http = require('http');
const { Buffer } = require('buffer');

const server = http.createServer((req, res) => {
    const hexData = req.url.substring(1);
    const ciphertext = Buffer.from(hexData, 'hex');
    const options = {
        hostname: 'motherload.td.org.uit.no',
        port: 8004,
        path: '',
        method: 'POST'
    };

    const request = http.request(options, response => {
        let data = '';
        response.on('data', chunk => {
            data += chunk;
        });
        response.on('end', () => {
            if (response.statusCode === 200) {
                console.log(hexData + " 200")
                res.writeHead(200, { 'Content-Type': 'text/plain' });
                res.end();
            } else {
                res.writeHead(400);
                console.log(hexData + " 400")
                res.end();
            }
        });
    });

    request.on('error', error => {
        res.writeHead(400);
        res.end();
    });

    request.write(ciphertext);
    request.end();
});

server.listen(8000, () => {
    console.log('Server running at http://localhost:8000/');
});

File: uithack-2023/crypto/oracle/middleman.py
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
from binascii import unhexlify
import requests

class SimpleHTTPRequestHandler(BaseHTTPRequestHandler):
    
    def do_GET(self):
        try:
            url_value = self.path[1:].encode()
            ciphertext = bytearray(unhexlify(url_value))

            response = requests.post("http://motherload.td.org.uit.no:8004", data=ciphertext)

            if response.status_code == 200:
                self.send_response(200)
                self.send_header('Content-type', 'text/plain')
                self.end_headers()
                self.wfile.write(b'')
            else:
                self.send_error(400, "")
            return
        except: 
            self.send_error(400, "")

def run(server_class=HTTPServer, handler_class=SimpleHTTPRequestHandler):
    server_address = ('', 8000)
    httpd = server_class(server_address, handler_class)
    print('Starting httpd...')
    httpd.serve_forever()

run()